### PR TITLE
Add order to analysis data

### PR DESCRIPTION
**Pull request by bittsitt**
Been a bit busy, but finally got around to this one. Address #1425.

I realized that we have to make a copy of the data since it can change at any time, making sorting difficult if we don't have a copy.  I added a class to store the data. Open to other suggestions.

### CODE REVIEW

File: src/UCTSearch.cpp [hunk 1]
```diff
@@ -40,6 +41,37 @@ using namespace Utils;
 
 constexpr int UCTSearch::UNLIMITED_PLAYOUTS;
 
+class OutputAnalysisData {
+public:
+    OutputAnalysisData(std::string move, int visits, int winrate, std::string pv) :
```
Comment (gcp) [inline src/UCTSearch.cpp]: const std::string&

File: src/UCTSearch.cpp [hunk 2]
```diff
@@ -40,6 +41,37 @@ using namespace Utils;
...
+                          " winrate " + std::to_string(m_winrate);
+        if (order >= 0) {
+            tmp += " order " + std::to_string(order);
+        }
+        tmp += " pv " + m_pv;
+        return tmp;
+    }
+
+    friend bool operator<(const OutputAnalysisData& a, const OutputAnalysisData& b)
+    {
```
Comment (gcp) [inline src/UCTSearch.cpp]: backet at the end of the previous line

File: src/UCTSearch.cpp [hunk 3]
```diff
@@ -40,6 +41,37 @@ using namespace Utils;
 
 constexpr int UCTSearch::UNLIMITED_PLAYOUTS;
 
+class OutputAnalysisData {
+public:
+    OutputAnalysisData(std::string move, int visits, int winrate, std::string pv) :
+        m_move(move), m_visits(visits), m_winrate(winrate), m_pv(pv) {};
+
+    std::string infoString(int order) const {
+        std::string tmp = "info move " + m_move + " visits " + std::to_string(m_visits) +
```
Comment (gcp) [inline src/UCTSearch.cpp]: auto tmp

Also I think the general style of the code would make this get_info_string().

File: src/UCTSearch.cpp [hunk 4]
```diff
@@ -236,6 +268,9 @@ void UCTSearch::dump_stats(FastState & state, UCTNode & parent) {
 }
 
 void UCTSearch::output_analysis(FastState & state, UCTNode & parent) {
+    // We need to make a copy of the data before sorting
+    std::vector<OutputAnalysisData> sortable_data;
```
Comment (gcp) [inline src/UCTSearch.cpp]: auto sortable_data =

File: src/UCTSearch.cpp [hunk 5]
```diff
@@ -251,13 +286,24 @@ void UCTSearch::output_analysis(FastState & state, UCTNode & parent) {
...
         auto move_eval = node->get_visits() ?
-            static_cast<int>(node->get_pure_eval(color) * 10000) : 0;
-        gtp_printf_raw("info %s %s %s %d %s %d %s %s\n",
-                       "move", move.c_str(),
-                       "visits", node->get_visits(),
-                       "winrate", move_eval,
-                       "pv", pv.c_str());
+                         static_cast<int>(node->get_pure_eval(color) * 10000) : 0;
+        // Store data in array
+        sortable_data.push_back(OutputAnalysisData(move, node->get_visits(), move_eval, pv));
```
Comment (gcp) [inline src/UCTSearch.cpp]: emplace_back and you don't need to mention the constructor then

File: src/UCTSearch.cpp [hunk 6]
```diff
@@ -251,13 +286,24 @@ void UCTSearch::output_analysis(FastState & state, UCTNode & parent) {
...
+
+    }
+    // Sort array to decide order
+    std::stable_sort(rbegin(sortable_data), rend(sortable_data));
+
+    int i = 0;
```
Comment (gcp) [inline src/UCTSearch.cpp]: auto

**Comment by bittsitt** [comment]
Should've remembered the auto... emplace_back was really neat!

**Pull request merged**